apps/src/blockly/addons/cdoBlockShadow.js

```diff
@@ -0,0 +1,141 @@
+      // If this block has itself as a parent and is in a flyout workspace,
+      // it may be a in a mini toolbox. We check the parentBlockId on the block to
+      // confirm it is still inside the parent flyout workspace.
+      if (workspace && workspace.isFlyout && block.parentBlockId) {
+        const potentialParent = Blockly.getMainWorkspace().getBlockById(
+          block.parentBlockId
+        );
+        // Confirm potential parent is the correct type and it
+        // does contain this block's current workspace.
+        if (potentialParent.type === blockShadowData.parent) {
+          const inputs = potentialParent.inputList.filter(
+            input => input.name === 'flyout_input'
+          );
+          if (inputs.length > 0) {
+            const flyoutWorkspace = inputs[0].fieldRow[0]?.flyout_?.workspace_;
+            // If tthis block's workspace id matches the flyout workspace id,
+            // then we have found the parent.
+            if (flyoutWorkspace && flyoutWorkspace.id === block.workspace.id) {
+              parent = potentialParent;
+            }
+          }
+        }
+      }
+    }
+  }
+  if (parent) {
+    return getImageUrlFromParent(parent, blockShadowData.imageIndex);
+  } else {
+    // The block is probably disconnected from any root or toolbox. Reset to
+    // default text.
+    return '';
+  }
+}
+
+// Given a shadow block and an optional parent block id, update the image displayed
+// on the shadow block to match the image selected in the parent block. See getShadowedBlockImageUrl
+// for details on how we find the parent block.
+export function updateShadowedBlockImage(block, parentId) {
+  const url = getShadowedBlockImageUrl(block, parentId);
+  changeShadowedImage(url, block);
+}
+
+// Get the image url from the parent block's inputList at the given index.
+// We find the image url by looking at the connection to the input at the given index.
+// If the parent does not have an image at the given index, return an empty string.
+function getImageUrlFromParent(parentBlock, imageIndexOnParent) {
+  const targetConnection =
+    parentBlock.inputList[imageIndexOnParent]?.connection?.targetConnection;
+  // We only want to get the image from a connection block that is not an insertion marker.
+  // If the block is an insertion marker that means the block is being dragged
+  // and is not yet connected to the parent block.
+  if (
+    targetConnection &&
+    targetConnection.sourceBlock_ &&
+    !targetConnection.sourceBlock_.isInsertionMarker()
+  ) {
+    return (
+      targetConnection.sourceBlock_.inputList[0]?.fieldRow[0]?.imageElement_?.getAttribute(
+        'xlink:href'
+      ) || ''
+    );
+  } else {
+    return '';
+  }
+}
+
+// Set the given childBlock's image input to the given imageUrl,
+// or reset the image input to the default long text if the imageUrl is empty.
+function changeShadowedImage(imageUrl, childBlock) {
```

Comment:
I'm curious why we went with `childBlock` instead of just `block` for this parameter. In theory, could this be used to set values on any appropriate block?

Reply:
just my inconsistency. I'll update to just be `block`.